docs/internal/foreign_code.md

```diff
@@ -62,32 +65,210 @@ We hope that the region abstraction will make programming with sandboxes natural
 
-This instructs the compiler to expose the C++ namespace and class structure the `LLVM` Verona class.
-C++ namespaces and classes will appear as nested classes within the Verona `LLVM` class.
-For example, `LLVM.llvm.SmallVector[LLVM.int, 12]` corresponds to `llvm::SmallVector<int, 12>`.
+### Overloaded operators
 
-Instantiating this class would require the user to specify a sandboxing policy, for example (again, straw-man syntax):
+Verona allows overloading most of the operators that C++ provides and so these can be trivially mapped just like any other methods.
+The main exceptions relate to assignment, though Verona's `apply` sugar provides similar functionality.
+Classes from C++ need to opt out of `create` and `apply` sugar because these reflect patterns that don't exist in C++ codebases and so it's possible that apply sugar could be applied for C++ assignment operators.
+
+### Casts
+
+C++ casts are exposed as generic functions on the class that represents the library.
+Any C library has a `$cast[T,U](U) : T` function defined, which takes any C built-in type or pointer type and returns any other.
+For C++ codebases, this performs a 'C-style cast', including invoking user-defined conversion operators.
+Any C++ library will also expose `static_cast[T,U](U) : T` and `reinterpret_cast[T,U](U) : T` functions and, if compiled with RTTI enabled, a `dynamic_cast[T,U](U) : T` function.
+These can all be used with any combination of types for which their C/C++ analogues are defined and with the same semantics.
+The Verona syntax implicitly applies to pointers when given any value except a C built-in type.
+C built-in types are passed as values to the cast functions unless wrapped in `Ref` cells.
+For example:
 
 ```verona
-var llvm = LLVM.Create(PROCESS);
+// Equivalent to auto *x = new SomeStruct();
+var x = C.SomeStruct();
+// Equivalent to auto *y = static_cast<SomeOtherStruct*>(x);
+var y = C.static_cast[C.SomeOtherStruct](x);
+var i : C.int = 12;
+// Equivalent to f = static_cast<float>(i);
+var f = C.static_cast[C.float](i);
+// Equivalent to float *f_type_punned = reinterpret_cast<float*>(&i);
+var f_type_punned : C.float = C.reinterpret_cast[Ref[C.float], Ref[C.int]](i);
 ```
 
-This would create a new process that loaded the LLVM library and had RPC invocations for any of the LLVM functions that this program called.
-The `llvm` object in this example is a new isolated region.
-The programmer can then construct LLVM types within that region.
+Note that the last line of this is explicitly violating type safety.
+All of these operations are restricted to values that are within a sandboxed foreign region where type safety is explicitly not guaranteed.
+
+### Exceptions
+
+**This doesn't make me very happy, better suggestions welcome**
+Every sandboxed function may throw a `SandboxError` exception if the sandbox has failed and so all invocations of sandboxed code must be wrapped in a `try` block.
```

Comment:
This may not be necessary when interfacing with languages that do not have unchecked exceptions... Are there any?

Reply:
Java / .NET, for example.  If we called gcj-compiled Java or AoT-compiled C# then we'd be able to get a complete list of all possible exceptions (the checked list plus runtime exceptions like `NullPointerDereference`).  

I can imagine wanting to build higher-level interfaces on top of the C/C++ interop layer, for example something talking to CPython to provide a direct Verona <-> Python bridge that doesn't require the Verona programmer to think about the CPython implementations, but that's very much a v2 feature.